pricegraph/src/orderbook.rs

```diff
@@ -156,45 +166,37 @@ impl Orderbook {
 
         let (base, quote) = (node_index(market.base), node_index(market.quote));
 
-        while let Err(NegativeCycle(predecessors, node)) =
-            bellman_ford::search(&self.projection, quote)
-        {
-            let path = path::find_cycle(&predecessors, node, Some(quote))
-                .expect("negative cycle not found after being detected");
-
-            let base_index_search = if path.first() == Some(&quote) {
-                path.iter().position(|node| *node == base)
-            } else {
-                // NOTE: If the path doesn't start with the quote token, don't
-                // even need to search for the base token.
-                None
-            };
-            let base_index = match base_index_search {
-                Some(value) => value,
-                None => {
-                    // NOTE: Skip negative cycles that are not along the
-                    // specified market.
-                    self.fill_path(&path).unwrap_or_else(|| {
-                        panic!(
-                            "failed to fill path along detected negative cycle {}",
-                            format_path(&path),
-                        )
-                    });
-
-                    continue;
+        while let Err(cycle) = bellman_ford::search(&self.projection, quote) {
+            let cycle_not_in_path = match cycle.with_starting_node(quote) {
+                Err(cycle) => cycle,
+                Ok(cycle) => {
+                    let base_index_search = cycle.0.iter().position(|node| *node == base);
```

Comment:
ubernit: Maybe we can implement the Iterable trait on cycle directly in order to avoid the .0 here.